### PR TITLE
Implement undelete and purge API for users

### PR DESCRIPTION
**Pull request by cganote**
Refactored the controller vs. manager for the deletion and purge of users. Purge now wipes out the username (regardless of gdrp setting) and takes out histories. Undelete should now work.

### CODE REVIEW

**Comment by martenson** [comment]
@galaxybot test this

**Comment by martenson** [comment]
@galaxybot test this

**Comment by hexylena** [comment]
> Purge now wipes out the username...and takes out histories.

:tada::heart::tada::heart::tada::heart:

**Review by cganote — COMMENTED** [review]
Great, thank you!

**Comment by martenson** [comment]
just now noticed https://github.com/galaxyproject/galaxy/pull/8299 :/

**Comment by martenson** [comment]
I cherry picked the test commits from https://github.com/galaxyproject/galaxy/pull/8299 and fixed them. This PR now supersedes the other one.

**Comment by martenson** [comment]
@galaxybot test this

**Comment by hexylena** [comment]
Fixes https://github.com/galaxyproject/galaxy/issues/930

File: lib/galaxy/managers/users.py [hunk 1]
```diff
@@ -109,10 +111,108 @@ def create(self, email=None, username=None, password=None, **kwargs):
...
+            raise exceptions.ConfigDoesNotAllowException('The configuration of this Galaxy instance does not allow admins to delete users.')
+        # Check first if it is already deleted. If so, say so and do nothing.
+        # Unsure if we should add username to the log.debug or if that breaks the GDPR support
+        if user.deleted:
+            log.debug("User already deleted, but setting delete flag just in case.")
+        if user.purged:
+            log.debug("Cannot delete purged user, but setting delete flag just in case.")
         user.deleted = True
         self.session().add(user)
         self.session().flush()
```
Comment (nsoranzo) [inline lib/galaxy/managers/users.py]: I'd just replace lines 117-125 with:
```python
        super(UserManager, self).delete(user)
```

File: lib/galaxy/managers/users.py [hunk 2]
```diff
@@ -109,10 +111,108 @@ def create(self, email=None, username=None, password=None, **kwargs):
...
+        """Remove the deleted flag for the given user."""
+        if not self.app.config.allow_user_deletion:
+            raise exceptions.ConfigDoesNotAllowException('The configuration of this Galaxy instance does not allow admins to undelete users.')
+        if user.purged:
+            raise exceptions.ItemDeletionException('Purged user cannot be undeleted.')
+        if not user.deleted:
+            log.debug('User \'%s\' has not been deleted, so it cannot be undeleted.' % user.email)
+        user.deleted = False
+        self.session().add(user)
+        self.session().flush()
```
Comment (nsoranzo) [inline lib/galaxy/managers/users.py]: I'd replace lines 133-137 with:
```python
        super(UserManager, self).undelete(user)
```

File: lib/galaxy/managers/users.py [hunk 3]
```diff
@@ -109,10 +111,108 @@ def create(self, email=None, username=None, password=None, **kwargs):
...
+
+    def purge(self, user):
+        """Purge the given user. They must have the deleted flag already."""
+        if not self.app.config.allow_user_deletion:
+            raise exceptions.ConfigDoesNotAllowException('The configuration of this Galaxy instance does not allow admins to delete or purge users.')
+        if not user.deleted:
+            raise exceptions.MessageException('User \'%s\' has not been deleted, so they cannot be purged.' % user.email)
```
Comment (nsoranzo) [inline lib/galaxy/managers/users.py:132]: I'd remove lines 143-144, purging without deleting first is fine by me.

File: lib/galaxy/managers/users.py [hunk 4]
```diff
@@ -109,10 +111,108 @@ def create(self, email=None, username=None, password=None, **kwargs):
...
+                addr.address = new_secure_hash(addr.address + pseudorandom_value)
+                addr.city = new_secure_hash(addr.city + pseudorandom_value)
+                addr.state = new_secure_hash(addr.state + pseudorandom_value)
+                addr.postal_code = new_secure_hash(addr.postal_code + pseudorandom_value)
+                addr.country = new_secure_hash(addr.country + pseudorandom_value)
+                addr.phone = new_secure_hash(addr.phone + pseudorandom_value)
+                self.session().add(addr)
+
+        self.session().add(user)
+        self.session().flush()
```
Comment (nsoranzo) [inline lib/galaxy/managers/users.py]: I'd replace lines 166 and 213-214 with:
```python
        super(UserManager, self).purge(user)
```

File: lib/galaxy/managers/users.py (hunk 3 above)
Comment (martenson) [reply lib/galaxy/managers/users.py:132]: I think it is a useful extra safety layer before a destructive function is called. @cganote was it meant like this?

Comment (cganote) [reply lib/galaxy/managers/users.py:132]: Yep I was hoping to have a little bit of a safety net

Comment (nsoranzo) [reply lib/galaxy/managers/users.py:132]: I don't think this check belongs to the manager.

**Comment by bernt-matthias** [comment]
Any chance to cover this old one by me https://github.com/galaxyproject/galaxy/issues/3843

**Comment by martenson** [comment]
@bernt-matthias I wouldn't tie it to this PR though

**Comment by bernt-matthias** [comment]
> I wouldn't tie it to this PR though

sure

File: lib/galaxy/managers/users.py (hunk 3 above)
Comment (martenson) [reply lib/galaxy/managers/users.py:132]: I think it is better here than in the API controller though. Where would you put it?

**Comment by martenson** [comment]
@galaxybot test this

**Comment by martenson** [comment]
@galaxybot test this

File: test/api/test_users.py [hunk 5]
```diff
@@ -74,6 +77,36 @@ def test_admin_update(self):
...
+        user = self._setup_user(TEST_USER_EMAIL_DELETE)
+        self._delete("users/%s" % user["id"], admin=True)
+        updated_user = self._get("users/deleted/%s" % user['id'], admin=True).json()
+        assert updated_user['deleted'] is True, updated_user
+
+    def test_purge_user(self):
+        """Delete user and then purge them."""
+        user = self._setup_user(TEST_USER_EMAIL_PURGE)
+        response = self._delete("users/%s" % user["id"], admin=True)
+        self._assert_status_code_is(response, 200)
```
Comment (nsoranzo) [inline test/api/test_users.py]:
```suggestion
        self._assert_status_code_is_ok(response)
```

File: test/api/test_users.py [hunk 6]
```diff
@@ -74,6 +77,36 @@ def test_admin_update(self):
...
+        data = dict(purge="True")
+        response = self._delete("users/%s" % user["id"], data=data, admin=True)
+        self._assert_status_code_is(response, 200)
```
Comment (nsoranzo) [inline test/api/test_users.py]:
```suggestion
        self._assert_status_code_is_ok(response)
```

File: lib/galaxy/managers/users.py (hunk 3 above)
Comment (nsoranzo) [reply lib/galaxy/managers/users.py:132]: In the UI (`lib/galaxy/webapps/galaxy/controllers/admin.py` ?), I don't think it's the API business to ask users for confirmation or similar.

File: lib/galaxy/managers/users.py [hunk 7]
```diff
@@ -109,7 +111,91 @@ def create(self, email=None, username=None, password=None, **kwargs):
...
+
+    def undelete(self, user):
+        """Remove the deleted flag for the given user."""
+        if not self.app.config.allow_user_deletion:
+            raise exceptions.ConfigDoesNotAllowException('The configuration of this Galaxy instance does not allow admins to undelete users.')
+        if user.purged:
+            raise exceptions.ItemDeletionException('Purged user cannot be undeleted.')
+        super(UserManager, self).undelete(user)
+
+    def purge(self, user):
```
Comment (nsoranzo) [inline lib/galaxy/managers/users.py]:
```suggestion
    def purge(self, user, flush=True):
```

File: lib/galaxy/managers/users.py [hunk 8]
```diff
@@ -109,7 +111,91 @@ def create(self, email=None, username=None, password=None, **kwargs):
         return user
 
     def delete(self, user):
```
Comment (nsoranzo) [inline lib/galaxy/managers/users.py]:
```suggestion
    def delete(self, user, flush=True):
```

File: lib/galaxy/managers/users.py [hunk 9]
```diff
@@ -109,7 +111,91 @@ def create(self, email=None, username=None, password=None, **kwargs):
...
-        user.deleted = True
+        """Mark the given user deleted."""
+        if not self.app.config.allow_user_deletion:
+            raise exceptions.ConfigDoesNotAllowException('The configuration of this Galaxy instance does not allow admins to delete users.')
+        super(UserManager, self).delete(user)
+
+    def undelete(self, user):
```
Comment (nsoranzo) [inline lib/galaxy/managers/users.py]:
```suggestion
    def undelete(self, user, flush=True):
```

File: lib/galaxy/managers/users.py [hunk 10]
```diff
@@ -109,7 +111,91 @@ def create(self, email=None, username=None, password=None, **kwargs):
         return user
 
     def delete(self, user):
-        user.deleted = True
+        """Mark the given user deleted."""
+        if not self.app.config.allow_user_deletion:
+            raise exceptions.ConfigDoesNotAllowException('The configuration of this Galaxy instance does not allow admins to delete users.')
+        super(UserManager, self).delete(user)
```
Comment (nsoranzo) [inline lib/galaxy/managers/users.py]:
```suggestion
        super(UserManager, self).delete(user, flush=flush)
```

File: lib/galaxy/managers/users.py [hunk 11]
```diff
@@ -109,7 +111,91 @@ def create(self, email=None, username=None, password=None, **kwargs):
...
+
+    def undelete(self, user):
+        """Remove the deleted flag for the given user."""
+        if not self.app.config.allow_user_deletion:
+            raise exceptions.ConfigDoesNotAllowException('The configuration of this Galaxy instance does not allow admins to undelete users.')
+        if user.purged:
+            raise exceptions.ItemDeletionException('Purged user cannot be undeleted.')
+        super(UserManager, self).undelete(user)
```
Comment (nsoranzo) [inline lib/galaxy/managers/users.py]:
```suggestion
        super(UserManager, self).undelete(user, flush=flush)
```

File: lib/galaxy/managers/users.py [hunk 12]
```diff
@@ -109,7 +111,91 @@ def create(self, email=None, username=None, password=None, **kwargs):
...
+                addr.institution = new_secure_hash(addr.institution + pseudorandom_value)
+                addr.address = new_secure_hash(addr.address + pseudorandom_value)
+                addr.city = new_secure_hash(addr.city + pseudorandom_value)
+                addr.state = new_secure_hash(addr.state + pseudorandom_value)
+                addr.postal_code = new_secure_hash(addr.postal_code + pseudorandom_value)
+                addr.country = new_secure_hash(addr.country + pseudorandom_value)
+                addr.phone = new_secure_hash(addr.phone + pseudorandom_value)
+                self.session().add(addr)
+        # Purge the user
+        user.purged = True
```
Comment (nsoranzo) [inline lib/galaxy/managers/users.py]:
```suggestion
        super(UserManager, self).purge(user, flush=flush)
```

File: lib/galaxy/managers/users.py [hunk 13]
```diff
@@ -109,7 +111,91 @@ def create(self, email=None, username=None, password=None, **kwargs):
...
         self.session().add(user)
```
Comment (nsoranzo) [inline lib/galaxy/managers/users.py]:
```suggestion
```

File: lib/galaxy/managers/users.py [hunk 14]
```diff
@@ -109,7 +111,91 @@ def create(self, email=None, username=None, password=None, **kwargs):
...
         self.session().flush()
```
Comment (nsoranzo) [inline lib/galaxy/managers/users.py]:
```suggestion
```

File: lib/galaxy/managers/users.py (hunk 3 above)
Comment (martenson) [reply lib/galaxy/managers/users.py:132]: This is not a confirmation, this is a built-in safety measure before a destructive operation happens. I think it is much aligned with how carefully Galaxy handles similar situations.

**Comment by martenson** [comment]
@nsoranzo I think I'm not a fan of going from api to manager then to mixin, then same manager again, then same mixin again for one method. It feels confusing so I'd like to avoid that.

File: lib/galaxy/managers/users.py (hunk 3 above)
Comment (nsoranzo) [reply lib/galaxy/managers/users.py:132]: History purge doesn't have this check in either the manager or the API.

Comment (martenson) [reply lib/galaxy/managers/users.py:132]: Purge history is a user-centric operation that has an extra ui confirmation. Purge user is an admin operation, usually way more destructive, and exposed through API. I do not think the extra confirmation belongs just to the UI. I am not opposed to adding a UI check, but this PR does a good job at trying to minimize impact of potential admin errors.

**Comment by nsoranzo** [comment]
> @nsoranzo I think I'm not a fan of going from api to manager then to mixin, then same manager again, then same mixin again for one method. It feels confusing so I'd like to avoid that.

I think these are good abstractions. For reference:
- the API takes care of the communication with the client, checks the inputs and permissions, then invokes the methods of the right manager
- the manager inherits and specialises some methods from superclasses and mixins, ensuring a consistent interface
- the `PurgableManagerMixin` mixin implements a generic `purge()` method, which just calls the `delete()` method and sets the `purged` attribute of the item
- for `UserManager`, we need to specialise both the `delete()` and `purge()` method, inherited from `PurgableManagerMixin`

**Comment by martenson** [comment]
I don't think the path manager>mixin>manager>mixin>done  is better than manager>done but at this point it seems like a matter of opinion and I will defer to yours.

**Review by nsoranzo — APPROVED** [review]

**Review by hexylena — APPROVED** [review]
Fine for me :)

**Pull request merged**